crypto/hpke/hpke.c

```diff
@@ -0,0 +1,3137 @@
+     */
+    if (EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_GET_TAG, taglen, tag) != 1) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    memcpy(ciphertext + ciphertextlen, tag, taglen);
+    ciphertextlen += taglen;
+    if (ciphertextlen > *cipherlen) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    *cipherlen = ciphertextlen;
+    memcpy(cipher, ciphertext, ciphertextlen);
+
+err:
+    EVP_CIPHER_CTX_free(ctx);
+    EVP_CIPHER_free(enc);
+    OPENSSL_free(ciphertext);
+    return erv;
+}
+
+/*
+ * @brief RFC5869 HKDF-Extract
+ *
+ * @param libctx is the context to use (normally NULL)
+ * @param suite is the ciphersuite
+ * @param mode5869 - controls labelling specifics
+ * @param salt - surprisingly this is the salt;-)
+ * @param saltlen - length of above
+ * @param label - label for separation
+ * @param labellen - length of above
+ * @param zz - the initial key material (IKM)
+ * @param zzlen - length of above
+ * @param secret - the result of extraction (allocated inside)
+ * @param secretlen - bufsize on input, used size on output
+ * @return 1 for good otherwise bad
+ *
+ * Mode can be:
+ * - OSSL_HPKE_5869_MODE_PURE meaning to ignore all the
+ *   HPKE-specific labelling and produce an output that's
+ *   RFC5869 compliant (useful for testing and maybe
+ *   more)
+ * - OSSL_HPKE_5869_MODE_KEM meaning to follow section 4.1
+ *   where the suite_id is used as:
+ *   concat("KEM", I2OSP(kem_id, 2))
+ * - OSSL_HPKE_5869_MODE_FULL meaning to follow section 5.1
+ *   where the suite_id is used as:
+ *     concat("HPKE", I2OSP(kem_id, 2),
+ *          I2OSP(kdf_id, 2), I2OSP(aead_id, 2))
+ *
+ * Isn't that a bit of a mess!
+ */
+static int hpke_extract(OSSL_LIB_CTX *libctx,
+                        const OSSL_HPKE_SUITE suite, const int mode5869,
+                        const unsigned char *salt, const size_t saltlen,
+                        const char *label, const size_t labellen,
+                        const unsigned char *ikm, const size_t ikmlen,
+                        unsigned char *secret, size_t *secretlen)
+{
+    EVP_KDF *kdf = NULL;
+    EVP_KDF_CTX *kctx = NULL;
+    OSSL_PARAM params[5], *p = params;
+    int mode = EVP_PKEY_HKDEF_MODE_EXTRACT_ONLY;
+    const char *mdname = NULL;
+    unsigned char labeled_ikmbuf[INT_MAXSIZE];
+    unsigned char *labeled_ikm = labeled_ikmbuf;
+    size_t labeled_ikmlen = 0;
+    int erv = 1;
+    size_t concat_offset = 0;
+    size_t lsecretlen = 0;
+    uint16_t kem_ind = 0;
+    uint16_t kdf_ind = 0;
+
+    /* Handle oddities of HPKE labels (or not) */
+    switch (mode5869) {
+
+    case OSSL_HPKE_5869_MODE_PURE:
+        labeled_ikmlen = ikmlen;
+        labeled_ikm = (unsigned char *)ikm;
+        break;
+
+    case OSSL_HPKE_5869_MODE_KEM:
+        concat_offset = 0;
+        memcpy(labeled_ikm, OSSL_HPKE_VERLABEL, strlen(OSSL_HPKE_VERLABEL));
+        concat_offset += strlen(OSSL_HPKE_VERLABEL);
+        if (concat_offset >= INT_MAXSIZE) {
```

Comment:
Isn't this check too late? We've already overflowed the buffer if this fails. Using WPACKET would remove this.

Reply:
I don't believe this 1st check is too late.